### PR TITLE
feat: schematics to add rules-engine on a component

### PR DESCRIPTION
**Pull request by fpaul-1A**

### CODE REVIEW

File: packages/@o3r/rules-engine/package.json [hunk 1]
```diff
@@ -103,6 +103,7 @@
     "@o3r/localization": "workspace:^",
     "@o3r/logger": "workspace:^",
     "@o3r/schematics": "workspace:^",
+    "@schematics/angular": "~15.2.0",
```
Comment (kpanot) [inline packages/@o3r/rules-engine/package.json:106]: should be in the `peerDep` as well

File: packages/@o3r/rules-engine/schematics/rules-engine-to-component/index.ts [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+  const generateFiles: Rule = (tree: Tree, _context: SchematicContext) => {
+    if (options.path && tree.exists(options.path)) {
+      const projectName = options.projectName || getLibraryNameFromPath(options.path) || '';
+      const imports = [
+        {
+          from: '@angular/core',
+          importNames: ['inject', 'OnInit', 'OnDestroy']
+        },
+        {
+          from: '@o3r/configuration',
```
Comment (kpanot) [inline packages/@o3r/rules-engine/schematics/rules-engine-to-component/index.ts:24]: Should be removed by #303

**Comment by kpanot** [comment]
Missing the following points:
- The `ng-add` schematic of `@o3r/rules-engine` should register itself as `defaultCollection` in the `angular.json`
- The `ng g component` should have a parameter to request component to be generated with rules-engine option (and ask the question if the option is not defined).
- The `ng-add` schematic of `@o3r/rules-engine` should set the default value of the `@o3r/core:component` schematic option, in the `angular.json` relative to the generation of the component with the rules-engine option

File: packages/@o3r/core/schematics/component/container/schema.json [hunk 3]
```diff
@@ -55,6 +55,12 @@
       "description": "Generate component context",
       "default": true
     },
+    "useRulesEngine": {
+      "type": "boolean",
+      "description": "Generate component with rules-engine",
+      "default": false,
```
Comment (matthieu-crouzet) [inline packages/@o3r/core/schematics/component/container/schema.json:61]: default should be null, no ?

File: packages/@o3r/core/schematics/component/schema.json [hunk 4]
```diff
@@ -58,6 +58,12 @@
       "default": null,
       "nullable": true
     },
+    "useRulesEngine": {
+      "type": "boolean",
+      "description": "Generate component with rules-engine",
+      "default": false,
```
Comment (matthieu-crouzet) [inline packages/@o3r/core/schematics/component/schema.json:64]: same here

File: packages/@o3r/rules-engine/schematics/rules-engine-to-component/schema.json [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+      "$default": {
+        "$source": "projectName"
+      }
+    },
+    "path": {
+      "type": "string",
+      "description": "Directory containing the component"
+    }
+  },
+  "additionalProperties": true
```
Comment (matthieu-crouzet) [inline packages/@o3r/rules-engine/schematics/rules-engine-to-component/schema.json:19]: shouldn't path be a required param ?

File: packages/@o3r/schematics/src/utility/ast.ts [hunk 6]
```diff
@@ -169,7 +169,7 @@ export const generateBlockStatementsFromString = (str: string) => {
   const sourceFile = ts.createSourceFile(
     'index.ts',
     `function func() {
-      ${str}
+      ${str.replace(/^\s*/g, '')}
```
Comment (kpanot) [inline packages/@o3r/schematics/src/utility/ast.ts:172]: `.trim()` would do the job as well :)

**Review by kpanot — APPROVED** [review]

File: packages/@o3r/schematics/src/utility/ast.ts (hunk 6 above)
Comment (fpaul-1A) [reply packages/@o3r/schematics/src/utility/ast.ts:172]: `trim` won't remove the spaces at the beginning of a line

File: packages/@o3r/rules-engine/schematics/rules-engine-to-component/index.ts [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        ts.ScriptTarget.ES2020,
+        true
+      );
+      const result = ts.transform(sourceFile, [
+        (ctx) => (rootNode: ts.Node) => {
+          const { factory } = ctx;
+          const visit = (node: ts.Node): ts.Node => {
+            if (ts.isClassDeclaration(node)) {
+              const implementsClauses = node.heritageClauses?.find((heritageClause) => heritageClause.token === ts.SyntaxKind.ImplementsKeyword);
+              const interfaceToImplements = generateImplementsExpressionWithTypeArguments('OnInit, OnDestroy');
```
Comment (cpaulve-1A) [inline packages/@o3r/rules-engine/schematics/rules-engine-to-component/index.ts:68]: you don't check the interface are not already in the file?

Comment (cpaulve-1A) [reply packages/@o3r/rules-engine/schematics/rules-engine-to-component/index.ts:68]: except if the deduplicateHeritageCauses also works on interface?

File: packages/@o3r/rules-engine/schematics/rules-engine-to-component/index.ts [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+              const deduplicateHeritageCauses = (heritageCauses: any[]) =>
+                heritageCauses.filter((h, i) =>
+                  !heritageCauses.slice(i + 1).some((h2) => h2.kind === h.kind && h2.expression.escapedText === h.expression.escapedText)
+                );
+              const newImplementsClauses = implementsClauses
+                ? factory.updateHeritageClause(implementsClauses, deduplicateHeritageCauses([...implementsClauses.types, ...interfaceToImplements]))
+                : factory.createHeritageClause(ts.SyntaxKind.ImplementsKeyword, [...interfaceToImplements]);
+
+              const heritageClauses: ts.HeritageClause[] = [...(node.heritageClauses || [])]
```
Comment (cpaulve-1A) [inline packages/@o3r/rules-engine/schematics/rules-engine-to-component/index.ts:78]:
```suggestion
              const heritageClauses: ts.HeritageClause[] = (node.heritageClauses || [])
```

File: packages/@o3r/rules-engine/schematics/rules-engine-to-component/index.ts [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+              const newModifiers = ([] as ts.ModifierLike[])
+                .concat(ts.getDecorators(node) || [])
+                .concat(ts.getModifiers(node) || []);
+
+              const propertiesToAdd = generateClassElementsFromString(`
+                private readonly componentName = computeConfigurationName('${node.name?.escapedText as string}', '${projectName}');
+                private rulesEngineService = inject(RulesEngineService, {optional: true});
+              `);
+
+              const newNgOnInit = getSimpleUpdatedMethod(node, factory, 'ngOnInit', generateBlockStatementsFromString(`
```
Comment (cpaulve-1A) [inline packages/@o3r/rules-engine/schematics/rules-engine-to-component/index.ts:91]: what happens if there is already an ngOnInit method with a code block inside?

File: packages/@o3r/schematics/src/utility/ast.ts [hunk 10]
```diff
@@ -225,3 +225,66 @@ export const sortClassElement = (classElement1: ts.ClassElement, classElement2:
...
+/**
+ * Transformer to be used to fix the string literals generated by creating a new SourceFile (for instance, using #generateClassElementsFromString)
+ *
+ * @param ctx
+ */
+export const fixStringLiterals: ts.TransformerFactory<ts.SourceFile> = (ctx) => (rootNode) => {
+  const { factory } = ctx;
+  const visit = (node: ts.Node): ts.Node => {
+    if (node.kind === ts.SyntaxKind.StringLiteral) {
+      return factory.createStringLiteral(node.getText().replace(/^["'](.*)["']$/, '$1'), true);
```
Comment (cpaulve-1A) [inline packages/@o3r/schematics/src/utility/ast.ts:238]: what happens if we have a string lie the following: "Test 'Hello' worlds". Is that a valid scenario for this method?

**Review by vscaiceanu-1a — APPROVED** [review]

File: packages/@o3r/schematics/src/utility/ast.ts (hunk 10 above)
Comment (fpaul-1A) [reply packages/@o3r/schematics/src/utility/ast.ts:238]: it'll be generated as `'Test \'Hello\' worlds'`

File: packages/@o3r/rules-engine/schematics/rules-engine-to-component/index.ts (hunk 7 above)
Comment (fpaul-1A) [reply packages/@o3r/rules-engine/schematics/rules-engine-to-component/index.ts:68]: yes the naming is a bit misleading but `heritageClause` contains both extends and implements

File: packages/@o3r/rules-engine/schematics/rules-engine-to-component/index.ts (hunk 8 above)
Comment (fpaul-1A) [reply packages/@o3r/rules-engine/schematics/rules-engine-to-component/index.ts]: `node.heritageClauses` is of type NodeArray so it needs to be casted to an actual array

**Review by kpanot — APPROVED** [review]

File: packages/@o3r/rules-engine/schematics/rules-engine-to-component/index.ts [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+import { applyToUpdateRecorder, Change } from '@schematics/angular/utility/change';
+import * as ts from 'typescript';
+import { NgGenerateRulesEngineToComponentSchematicsSchema } from './schema';
+
+/**
+ * Generate the code to enable rules-engine on a component
+ *
+ * @param options
+ */
+export function ngGenerateRulesEngineToComponent(options: NgGenerateRulesEngineToComponentSchematicsSchema): Rule {
```
Comment (matthieu-crouzet) [inline packages/@o3r/rules-engine/schematics/rules-engine-to-component/index.ts:22]: You should add a check to throw if the component already have the rules engine

Comment (fpaul-1A) [reply packages/@o3r/rules-engine/schematics/rules-engine-to-component/index.ts:51]: done

**Review by kpanot — APPROVED** [review]

**Review by mrednic-1A — APPROVED** [review]

**Pull request merged**